### PR TITLE
support TF se_e2_a serialization; add a common test fixture to compare TF, PT, and DP models

### PR DESCRIPTION
**Pull request by njzjz**
Some codes are split from #2987.

### CODE REVIEW

File: source/tests/consistent/common.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        run_sess(sess, tf.global_variables_initializer())
+        return run_sess(
+            sess,
+            t_out_indentity,
+            feed_dict=feed_dict,
+        )
+
+    def test_tf_dp_consistent(self):
+        """Test whether TF and DP are consistent."""
+        if self.skip_dp:
```
Comment (iProzd) [inline source/tests/consistent/common.py:146]: if self.skip_dp or self.skip_tf?

**Review by iProzd — COMMENTED** [review]
Should the status of pt, tf, and dp in testing be considered equivalent? Should we set self.skip_tf, self.skip_pt, and self.skip_dp all? In future model implementations, there might be parts that are not available in tf as well.

File: source/tests/consistent/test_se_e2_a.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+            dtype=GLOBAL_NP_FLOAT_PRECISION,
+        )
+        self.atype = np.array([0, 1, 1, 0, 1, 1], dtype=np.int32)
+        self.box = np.array(
+            [13.0, 0.0, 0.0, 0.0, 13.0, 0.0, 0.0, 0.0, 13.0],
+            dtype=GLOBAL_NP_FLOAT_PRECISION,
+        )
+        self.natoms = np.array([6, 6, 2, 4], dtype=np.int32)
+
+    def build_tf(self, obj: Any, suffix: str) -> Tuple[list, dict]:
```
Comment (iProzd) [inline source/tests/consistent/test_se_e2_a.py:83]: This method should be in common.py? Most of descriptors may be the same.

File: source/tests/consistent/test_se_e2_a.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        )
+        return [t_des], {
+            t_coord: self.coords,
+            t_type: self.atype,
+            t_natoms: self.natoms,
+            t_box: self.box,
+            t_mesh: make_default_mesh(True, False),
+        }
+
+    def eval_dp(self, dp_obj: Any) -> Any:
```
Comment (iProzd) [inline source/tests/consistent/test_se_e2_a.py:106]: Maybe `eval_dp`, `eval_pt` and `compare_tf_dp_ret` should be in common.py as well?

Comment (wanghan-iapcm) [inline source/tests/consistent/test_se_e2_a.py:106]: would it better to explictly expose all inputs to developers? prevent using `self.coord`, `self.atype` and so on

File: source/tests/consistent/common.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        data2 = dp_obj.serialize()
+        dp_obj2 = self.dp_class.deserialize(data2)
+        ret3 = self.eval_dp(dp_obj2)
+        for rr2, rr3 in zip(ret2, ret3):
+            if isinstance(rr2, np.ndarray) and isinstance(rr3, np.ndarray):
+                np.testing.assert_allclose(rr2, rr3)
+            else:
+                self.assertEqual(rr2, rr3)
+
+    def test_tf_pt_consistent(self):
```
Comment (wanghan-iapcm) [inline source/tests/consistent/common.py:216]: should'nt the cross backend test be designed as test_dp_xxx_consistent?

**Review by wanghan-iapcm — COMMENTED** [review]
I am not sure if the current design is a good idea, because to test the implementation of pt, one have to install the tf. suppose we have 5 backends, the env would be very heavy for the developers. 

Ideally, the dp backend can be tested without any deep learning platforms, while any implementation, e.g. pt, can be tested without any other backend.

File: source/tests/consistent/test_se_e2_a.py (hunk 3 above)
Comment (njzjz) [reply source/tests/consistent/test_se_e2_a.py:106]: I don't think a descriptor and a fitting net can share the input. One may also want to test additional inputs.

File: source/tests/consistent/common.py (hunk 4 above)
Comment (njzjz) [reply source/tests/consistent/common.py]: After refactoring, the reference backend is dynamic, according to whether they are available.

https://github.com/deepmodeling/deepmd-kit/blob/a3b7e0f2d778892f101fe3f8888cfc89ca255eb5/source/tests/consistent/common.py#L191-L198

File: source/tests/consistent/test_se_e2_a.py (hunk 3 above)
Comment (njzjz) [reply source/tests/consistent/test_se_e2_a.py:77]: This class is not in common.py

File: source/tests/consistent/common.py [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+        data = obj.serialize()
+        return ret, data
+
+    def get_dp_ret_serialization_from_cls(self, obj):
+        ret = self.eval_dp(obj)
+        data = obj.serialize()
+        return ret, data
+
+    def get_reference_backend(self):
+        if not self.skip_dp:
```
Comment (wanghan-iapcm) [inline source/tests/consistent/common.py:221]:
```suggestion
    def get_reference_backend(self):
        """order of checking for ref: dp, tf, dp""" 
        if not self.skip_dp:
```

File: source/tests/consistent/common.py [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+            else:
+                self.assertEqual(rr1, rr2)
+
+    def tearDown(self) -> None:
+        """Clear the TF session."""
+        if not self.skip_tf:
+            clear_session()
+
+
+class DescriptorTest:
```
Comment (wanghan-iapcm) [inline source/tests/consistent/common.py:343]: provide in a separate file for descriptors? or shall we have a folder structure for the tests of different components?

**Review by wanghan-iapcm — COMMENTED** [review]
would it be possible to loop over a list of possible model parameters like 
https://github.com/deepmodeling/deepmd-kit/blob/392b9e0c5e44c1f1f93b2db2e63fa36ebcfb7914/source/tests/pt/model/test_se_e2_a.py#L43-L45

File: source/tests/consistent/common.py (hunk 6 above)
Comment (njzjz) [reply source/tests/consistent/common.py:343]: A folder structure is excellent.

File: source/tests/consistent/common.py (hunk 5 above)
Comment (wanghan-iapcm) [reply source/tests/consistent/common.py]: sry a typo

**Review by wanghan-iapcm — APPROVED** [review]
I like the design of decorator `parameterized`  very much.

**Review by iProzd — APPROVED** [review]

**Pull request merged**